### PR TITLE
Market and Verifreg: Deduplicate reads

### PR DESCRIPTION
**Pull request by aarshkshah1992**
For #278 .

### CODE REVIEW

File: actors/builtin/multisig/multisig_actor.go [hunk 1]
```diff
@@ -197,10 +203,16 @@ func (a Actor) Approve(rt vmr.Runtime, params *TxnIDParams) *ApproveReturn {
...
+		ptx, err := adt.AsMap(adt.AsStore(rt), st.PendingTxns)
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load pending transactions")
+
 		if !isSigner(rt.ResolveAddress, &st, callerAddr) {
 			rt.Abortf(exitcode.ErrForbidden, "%s is not a signer", callerAddr)
 		}
-		txn = a.getTransaction(rt, st, params.ID, params.ProposalHash, true)
+		txn = a.getTransaction(rt, st, ptx, params.ID, params.ProposalHash, true)
+
+		st.PendingTxns, err = ptx.Root()
```
Comment (anorth) [inline actors/builtin/multisig/multisig_actor.go:214]: There has been no mutation before this, the write is unnecessary.

File: actors/builtin/multisig/multisig_actor.go [hunk 2]
```diff
@@ -385,23 +403,27 @@ func (a Actor) approveTransaction(rt vmr.Runtime, txnID TxnID, txn *Transaction)
...
 }
 
-func (a Actor) getTransaction(rt vmr.Runtime, st State, txnID TxnID, proposalHash []byte, checkHash bool) *Transaction {
+func (a Actor) getTransaction(rt vmr.Runtime, st State, ptx *adt.Map, txnID TxnID, proposalHash []byte, checkHash bool) *Transaction {
 	var txn Transaction
 
 	// get transaction from the state trie
 	var err error
-	txn, err = st.getPendingTransaction(adt.AsStore(rt), txnID)
+	txn, err = getPendingTransaction(ptx, txnID, st.PendingTxns)
```
Comment (anorth) [inline actors/builtin/multisig/multisig_actor.go:426]: This is very confusing. Isn't `ptx` the same map as `st.PendingTxns`? But possibly one of them is stale now? `st` should not be a parameter if we're passing the adts (and also !!! it wasn't a pointer)

File: actors/builtin/multisig/multisig_state.go [hunk 3]
```diff
@@ -2,13 +2,11 @@ package multisig
...
 	address "github.com/filecoin-project/go-address"
-	cid "github.com/ipfs/go-cid"
-	errors "github.com/pkg/errors"
-	xerrors "golang.org/x/xerrors"
-
 	abi "github.com/filecoin-project/specs-actors/actors/abi"
 	big "github.com/filecoin-project/specs-actors/actors/abi/big"
 	adt "github.com/filecoin-project/specs-actors/actors/util/adt"
+	cid "github.com/ipfs/go-cid"
+	errors "github.com/pkg/errors"
```
Comment (anorth) [inline actors/builtin/multisig/multisig_state.go:9]: Fix import ordering/grouping

File: actors/builtin/multisig/multisig_state.go [hunk 4]
```diff
@@ -55,57 +53,15 @@ func (st *State) assertAvailable(currBalance abi.TokenAmount, amountToSpend abi.
...
+func getPendingTransaction(ptx *adt.Map, txnID TxnID, cid cid.Cid) (Transaction, error) {
 	var out Transaction
-	found, err := hm.Get(txnID, &out)
+	found, err := ptx.Get(txnID, &out)
 	if err != nil {
 		return Transaction{}, errors.Wrapf(err, "failed to read transaction")
 	}
 	if !found {
-		return Transaction{}, errors.Errorf("failed to find transaction %v in HAMT %s", txnID, as.PendingTxns)
+		return Transaction{}, errors.Errorf("failed to find transaction %v in HAMT %s", txnID, cid)
```
Comment (anorth) [inline actors/builtin/multisig/multisig_state.go:63]: Don't take the CID parameter just for this message. Actually using it for something meaningful would probably be a bug.

**Comment by aarshkshah1992** [comment]
Closed in favour of #715 .

**Pull request closed without merging**